src/core/meshcop/joiner.cpp

```diff
@@ -187,117 +228,154 @@ void Joiner::HandleDiscoverResult(otActiveScanResult *aResult, void *aContext)
+    }
+    else
+    {
+        mJoinerRouterIndex = 0;
+        TryNextJoinerRouter(OT_ERROR_NONE);
+    }
 
-        otLogDebgMeshCoP("Received Discovery Response (%s)",
-                         static_cast<Mac::ExtAddress &>(aResult->mExtAddress).ToString().AsCString());
+exit:
+    return;
+}
 
-        // Joining is disabled if the Steering Data is not included
-        if (aResult->mSteeringData.mLength == 0)
+void Joiner::SaveDiscoveredJoinerRouter(const otActiveScanResult &aResult)
+{
+    uint8_t       priority;
+    bool          doesAllowAny = true;
+    JoinerRouter *end          = OT_ARRAY_END(mJoinerRouters);
+    JoinerRouter *entry;
+    JoinerRouter *last;
+
+    // Check whether Steering Data allows any joiner (it is all 0xff).
+    for (uint8_t i = 0; i < aResult.mSteeringData.mLength; i++)
+    {
+        if (aResult.mSteeringData.m8[i] != 0xff)
         {
-            otLogDebgMeshCoP("No steering data, joining disabled");
-            ExitNow();
+            doesAllowAny = false;
+            break;
         }
+    }
 
-        SteeringDataTlv steeringData;
-        steeringData.SetLength(aResult->mSteeringData.mLength);
-        memcpy(steeringData.GetValue(), aResult->mSteeringData.m8, steeringData.GetLength());
+    otLogInfoMeshCoP("Joiner discover network: %s, pan:0x%04x, port:%d, chan:%d, rssi:%d, allow-any:%s",
+                     static_cast<const Mac::ExtAddress &>(aResult.mExtAddress).ToString().AsCString(), aResult.mPanId,
+                     aResult.mJoinerUdpPort, aResult.mChannel, aResult.mRssi, doesAllowAny ? "yes" : "no");
 
-        if (!steeringData.GetBit(mCcitt % steeringData.GetNumBits()) ||
-            !steeringData.GetBit(mAnsi % steeringData.GetNumBits()))
-        {
-            otLogDebgMeshCoP("Steering data does not include this device");
-            ExitNow();
-        }
+    priority = CalculatePriority(aResult.mRssi, doesAllowAny);
 
-        joinerRouter.mPriority = static_cast<uint16_t>(aResult->mRssi) + 0x80;
+    // We keep the list sorted based on priority. Find the place to
+    // add the new result.
 
-        if (!steeringData.DoesAllowAny())
+    for (entry = &mJoinerRouters[0]; entry < end; entry++)
+    {
+        if (priority > entry->mPriority)
         {
-            joinerRouter.mPriority += kSpecificPriorityBonus;
+            break;
         }
-
-        joinerRouter.mJoinerUdpPort = aResult->mJoinerUdpPort;
-        joinerRouter.mPanId         = aResult->mPanId;
-        joinerRouter.mChannel       = aResult->mChannel;
-        joinerRouter.mExtAddr       = static_cast<Mac::ExtAddress &>(aResult->mExtAddress);
-        AddJoinerRouter(joinerRouter);
     }
-    else
-    {
-        otError error = TryNextJoin();
 
-        if (error != OT_ERROR_NONE)
+    VerifyOrExit(entry < end);
+
+    // Find the last in-use element in the list after entry
+    // (in-use elements have non-zero priority).
+
+    for (last = entry; last < end; last++)
```

Comment:
If the goal is to minimize code, you could avoid searching for last and move the entire remainder of the list (which currently defaults to 2 entries).

Reply:
I agree... should be ok and would save us a bunch of lines.
Changed this in new pushed commit (also squashed the logic to keep the list sorted into the main commit).